### PR TITLE
Experimental TFLite fix for Whisper

### LINKED ISSUE #19691: Not able to run evaluate on whisper.tflite that got generated from TFWhisper model
### Model description

@gante
Generated HF TFwhisper model into whisper.tfllite model. However, I'm not sure how to evaluate the created whisper tflite model.

https://colab.research.google.com/github/usefulsensors/openai-whisper/blob/main/tflite_from_huggingface_whisper.ipynb

I would appreciate your assistance in evaluating whisper.tflite. The notebook mentioned above produces a whisper.tflite file.

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

_No response_

### LINKED ISSUE #32125: TF Lite model created from TFWhisperForConditionalGeneration.from_pretrained craches
### System Info

The main branch still does not include the fix:
https://github.com/huggingface/transformers/issues/19691#issuecomment-1791869884

namely
```
    #new_scores = tf.ones_like(scores, dtype=scores.dtype) * -float("inf")
    new_scores = tf.ones_like(scores, dtype=scores.dtype) * -float(1)
```

which causes TFLite Whisper model to crash when used as:
```
    interpreter = tf.lite.Interpreter(tflite_model_path)
    tflite_generate = interpreter.get_signature_runner()
    output = tflite_generate(input_features=input_features)
```

### Who can help?

 @ArthurZucker

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```
class GenerateModel(tf.Module):
    def __init__(self, model, forced_decoder_ids):
        super(GenerateModel, self).__init__()
        self.model = model
        self.forced_decoder_ids = forced_decoder_ids

    @tf.function(
        # shouldn't need static batch size, but throws exception without it (needs to be fixed)
        input_signature=[
            tf.TensorSpec((1, 80, 3000), tf.float32, name="input_features"),
        ],
    )
    def serving(self, input_features):
        outputs = self.model.generate(
            input_features,
            forced_decoder_ids=self.forced_decoder_ids,
            #max_new_tokens=223,  # change as needed
            return_dict_in_generate=True,
        )
        return {"sequences": outputs["sequences"]}

model = TFWhisperForConditionalGeneration.from_pretrained(model_name)
generate_model = GenerateModel(model=model, forced_decoder_ids=forced_decoder_ids)
tf.saved_model.save(generate_model, saved_model_dir, signatures={"serving_default": generate_model.serving})

# Convert the model
converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir)
converter.target_spec.supported_ops = [
        tf.lite.OpsSet.TFLITE_BUILTINS,  # enable TensorFlow Lite ops.
        tf.lite.OpsSet.SELECT_TF_OPS  # enable TensorFlow ops.
]

converter.optimizations = [tf.lite.Optimize.DEFAULT]
tflite_model = converter.convert()

# Save the model
with open(tflite_model_path, 'wb') as f:
        f.write(tflite_model)

....
    interpreter = tf.lite.Interpreter(tflite_model_path)
    tflite_generate = interpreter.get_signature_runner()
    output = tflite_generate(input_features=input_features)
```

causes crache on tflite_generate.

Applying the patch https://github.com/huggingface/transformers/issues/19691#issuecomment-1791869884 solves the problem

### Expected behavior

The code below should work without patch

```
class GenerateModel(tf.Module):
    def __init__(self, model, forced_decoder_ids):
        super(GenerateModel, self).__init__()
        self.model = model
        self.forced_decoder_ids = forced_decoder_ids

    @tf.function(
        # shouldn't need static batch size, but throws exception without it (needs to be fixed)
        input_signature=[
            tf.TensorSpec((1, 80, 3000), tf.float32, name="input_features"),
        ],
    )
    def serving(self, input_features):
        outputs = self.model.generate(
            input_features,
            forced_decoder_ids=self.forced_decoder_ids,
            #max_new_tokens=223,  # change as needed
            return_dict_in_generate=True,
        )
        return {"sequences": outputs["sequences"]}

model = TFWhisperForConditionalGeneration.from_pretrained(model_name)
generate_model = GenerateModel(model=model, forced_decoder_ids=forced_decoder_ids)
tf.saved_model.save(generate_model, saved_model_dir, signatures={"serving_default": generate_model.serving})

# Convert the model
converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir)
converter.target_spec.supported_ops = [
        tf.lite.OpsSet.TFLITE_BUILTINS,  # enable TensorFlow Lite ops.
        tf.lite.OpsSet.SELECT_TF_OPS  # enable TensorFlow ops.
]

converter.optimizations = [tf.lite.Optimize.DEFAULT]
tflite_model = converter.convert()

# Save the model
with open(tflite_model_path, 'wb') as f:
        f.write(tflite_model)

....
    interpreter = tf.lite.Interpreter(tflite_model_path)
    tflite_generate = interpreter.get_signature_runner()
    output = tflite_generate(input_features=input_features)
```

### PR DESCRIPTION
**Pull request by Rocketknight1**
Don't merge, this will probably break some tests!

Fixes #32125 
Fixes #19691

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Hi @sanchit-gandhi @amyeroberts, this PR fixes an issue with TFLite conversion of TFWhisper models. However, the cause of the bug was that our forced decoding logits processor sets unwanted logits to `-inf`, and that breaks TFLite. This PR uses `-10000.` instead, which should be within the `float16` representable range.

The output of the sampling before and after this PR should be the same, because a logit of `-10000` will yield a probability of 0 after softmax at any precision. However, it does create a small numerical difference in the logits between Torch and TF. This doesn't seem to break any of our cross-tests, though, even when I ran the slow ones locally.

As such, I think this is safe to merge, and I'll double-check with @dsame that it fixes the issue. Let me know if you're opposed, though!

File: src/transformers/generation/tf_logits_process.py [hunk 1]
```diff
@@ -569,7 +569,7 @@ def _force_token(generation_idx):
             batch_size = scores.shape[0]
             current_token = self.force_token_array[generation_idx]
 
-            new_scores = tf.ones_like(scores, dtype=scores.dtype) * -float("inf")
```
Comment (sanchit-gandhi) [inline src/transformers/generation/tf_logits_process.py:572]: Is it possible to take the min value for a given dtype, e.g. as we do in torch with:
```py
min_dtype = torch.finfo(dtype).min
```

**Review by sanchit-gandhi — COMMENTED** [review]
We've managed to refactor the PT Whisper generation code to completely remove the dependence on forced decoder ids -> this could be a nice task for a community contributor to port TF!

File: src/transformers/generation/tf_logits_process.py (hunk 1 above)
Comment (amyeroberts) [reply src/transformers/generation/tf_logits_process.py:572]: +1

**Review by amyeroberts — COMMENTED** [review]
Thanks for tackling this! 

For slow tests - to make sure they're all good with the CI, we can run on the CI runners by:
* Adding the `run-slow` label to the PR
* Pushing a commit with the message `[run_slow] whisper`

**Comment by Rocketknight1** [comment]
Ah! Given @sanchit-gandhi's comment, we probably shouldn't merge this PR - instead we should try to refactor the TF generation to not use forced decoder IDs, which avoids the problem entirely.
